crux_kv/src/lib.rs

```diff
@@ -2,36 +2,68 @@
 //!
 //! `crux_kv` allows Crux apps to store and retrieve arbitrary data by asking the Shell to
 //! persist the data using platform native capabilities (e.g. disk or web localStorage)
-//!
-//! This is still work in progress and extremely basic.
+
+pub mod error;
+
 use crux_core::capability::{CapabilityContext, Operation};
 use crux_core::macros::Capability;
+use error::KeyValueError;
 use serde::{Deserialize, Serialize};
 
 /// Supported operations
 #[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
-pub enum KeyValueOperation {
+pub enum KeyValueRequest {
     /// Read bytes stored under a key
-    Read(String),
+    Get { key: String },
     /// Write bytes under a key
-    Write(String, Vec<u8>),
+    Set { key: String, value: Vec<u8> },
+    /// Remove a key and its value
+    Delete { key: String },
+    /// Test if a key exists
+    Exists { key: String },
+}
+
+/// The result of reading from the key value store.
+///
+/// Note: we can't use `Result` and `Option` here because generics are not currently
+/// supported across the FFI boundary, when using the builtin typegen.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum KeyValueReadResult {
+    /// the value stored under the key, may be empty
+    Data { value: Vec<u8> },
+    /// whether the key exists or not
+    Exists { value: bool },
+    /// an error occurred
+    Err { error: KeyValueError },
+}
```

Comment:
I suppose you could merge the Read and Write result values into a single enum, and then it would 

```rust
/// this is the capability's output
KeyValueResult {
     Ok(KeyValueResponse),
     Err(KeyValueError),
}

KeyValueResponse {
    Data { value: ... },
    Updated { previous: ... }, // assume this handles Set and Delete
    Exists { present: ... },
}
```

not sure if that's better... 🤔 Less nesting, more invalid states, but we already do asserts on the response...?

Reply:
I like this better. Will update

Reply:
see https://github.com/redbadger/crux/pull/227/commits/3faa22baa18e506af63587ad834ac8edf6eead42